### DISCUSSION TITLE
How does one contribute?

### DISCUSSION THREAD

**Discussion by cbuteau** [plantuml/plantuml#1346] [General]
Hello,

NOT A BUG

I use plantuml embedded in markdown where I work and I love it.
I would love to debug and fix bugs or features for it on my spare time.

Where is the signup sheet? 
or 
Where is the process?

Thanks.

**Comment by arnaudroques** [comment]
Many thanks for your support !
There are no precise process.
You can contribute many ways:

- Find bugs or post enhancement suggestion ( http://plantuml.com/qa )
- Write some library ( http://plantuml.com/stdlib )
- Write or enhance some existing plugin ( http://plantuml.com/running )
- Contribute to the documentation ( http://wiki.plantuml.net/site/index )
- Clone the core library and propose a merge request. Even if the code under github is a a copy of the master repository, we accept github merge request (for example https://github.com/plantuml/plantuml/pull/196 ).
- If you have a good knowledge of Java and C language, you may review the port of GraphViz from Java to C ( http://plantuml.com/smetana02 )

It's really up to you!

**Comment by cbuteau** [comment]
Wow great response...
Who do I contact for the crazy port of GraphViz and C project?
Why C?
It seems if PlantUML wants to become internet worthy...it should do javascript.

Or which repository should I clone and run a demo for?

**Comment by VladimirAlexiev** [comment]
@cbuteau Smetana is a port of Graphviz from C to Java (and some part in JS?).
So to work on this port, you should know C reasonably well.

**Reply by cbuteau** [reply]
Define resonably well.
I have worked on some key product related C++ work and when I was younger worked on a mostly C based communication library and system.

I code javascript everyday now but I spent a good period of my career with C# so the java ramp up should be easy.

Where do I get the source?
If I navigate to here.
https://github.com/plantuml/smetana/tree/V2_38

Cloning the URL the Code button gives me gives me nothing.

So where do I get the source?
How do I compile and run it?
Where are the bugs?

Thanks for all the positive replies.
-Chris

**Reply by arnaudroques** [reply]
This part is really a hard one.

First, the code is now on [on a branch](https://github.com/plantuml/smetana/tree/legacy). But all this is legacy.

The translation C-to-Java was somehow a success, so [last version of the code is now integrated into plantuml itself](https://github.com/plantuml/plantuml/tree/master/src/gen). 

So we keep [this repo](https://github.com/plantuml/smetana) for historical reason because it shows how we convert C code to Java Code.
**This repo is really useless now.**

However, it does _not_ mean that the work is finished.
Even if it is working, there are still some issues.

And we did not explain the **real final goal**, which is to refactor this code so that it looks like regular Java code.
Because we think that we could do much more with graphviz algorithms than the dot language currently allows.

So if you are still interested, we could work together:
- we could set up a "smetana" branch on plantuml project. In that branch, we could remove all plantuml code and keep only Graphviz code (that is smetana code)
- we definitively need to add documentation and tests there. Maybe you could start by this?

Keep in mind that this is really a very unusual Java project. It's a good point that you know C.
Just to give you an idea, you can have a look at [this file](https://github.com/plantuml/plantuml/blob/master/src/gen/lib/cdt/dttree__c.java#L110). It is Java code, but it really looks like C code.

Still interested?
If you are, you might go to the [discord server](https://discord.gg/sXhzexAQGh), on the `#smetana` room.

**Reply by VladimirAlexiev** [reply]
@arnaudroques Please Archive the https://github.com/plantuml/smetana repo.
And I posted two PRs to add links to 2 READMEs.
While you are at it, please add
> the [discord server](https://discord.gg/sXhzexAQGh), on the #smetana room

(sorry I didn't see this link!)

**Comment by cbuteau** [comment]
Wow I mean wow
It looks like you ported pointer magic into Java
That must have been hard.

So I am still interested
But I don't want to make you create a branch if I am not going to commit.
So I need to think on this.

The Java EcoSystem is similar to the C# one in that you find libraries to build what you need.
It sounds like you should almost start from the top down with the goals and algorithms.

But you are talking starting with interfaces testing and documenting what the underlying mechanisms are first?

How is it incorporated in plantuml?
Is the c dependency gone?

So from the top level?
What are the main tasks that PlantUML uses from this library?
What are the inputs and the outputs?

**Reply by VladimirAlexiev** [reply]
I worked  a bit on graphviz 25 years ago since I needed it for graphs in my PhD.
I added arrowhead types and ports (where two arrows land on the same spot on the node).

PlantUML uses Graphviz to do the graph layout. 
This includes node positioning and edge routing.

It doesn't use it to draw the nodes, but maybe uses it to clip edges and position arrowheads on the node boundary.

Input is probably the "dot" text format (given that PlantUML can take that as input: https://plantuml.com/dot).

Output is node positions, and edge splines/polylines. 
Maybe in https://graphviz.org/docs/outputs/canon/ text format, or as Java objects.

**Comment by boessu** [comment]
Hello all

I just add the question here as it seems for me the right place to ask: I've contributed the edgy-library in https://github.com/plantuml/plantuml-stdlib as a push, following the description in CONTRIBUTING.md. However, it seems as there is a compression-area in https://github.com/plantuml/plantuml/tree/master/stdlib where the library has to been included (if I understood that part in the right way).

1. Do I have to fork plantuml too and add there a compressed form of the library too as a pull-request or is this a part of the release workflow to include the libraries in https://github.com/plantuml/plantuml-stdlib into https://github.com/plantuml/plantuml/tree/master/stdlib?
2. If I have to contribute that: The linked file in https://github.com/plantuml/plantuml/blob/master/src/net/sourceforge/plantuml/preproc/Stdlib.java seems to explain how to read a compressed file, but not how to build that. How do I compress my library?

**Reply by The-Lum** [reply]
Hello @boessu and all, 

Don't worry ;
That is not yet possible by contributors ;
That is a manualy action on the workflow (by @arnaudroques ) and not yet an automated action.

As all the commits on this plantuml/stdlib directory:
- https://github.com/plantuml/plantuml/commit/a66215a373cd26ca2bb548a6cf26ecba38d9017f

_If that can help,_
_Awaiting the @arnaudroques actions,_

_[During this period, have an Happy New Year ✨️]_

Regards,
Th.

**Reply by arnaudroques** [reply]
We have just pushed the edgy tools [in the packaged standard library](https://github.com/plantuml/plantuml/commit/533c27cba52db66726a924fa26913a602478a3f9).

We must add some info in [CONTRIBUTING.md](https://github.com/plantuml/plantuml-stdlib/blob/master/CONTRIBUTING.md) about this needed manual action. So your contribution is the first of 2024!

Additionally, we should clearly outline the file compression method used. Implementing a GitHub action for this process will help automate it, ensuring efficiency and consistency.

P.S.: Wishing everyone a joyous and prosperous New Year!
